### PR TITLE
Add new resource turbot_file.

### LINKED ISSUE #34: Add new resource turbot_file

### PR DESCRIPTION
**Pull request by RupeshPatil20**
Closes #34

### CODE REVIEW

File: turbot/resource_turbot_file.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"data": {
+				Type:             schema.TypeString,
+				Required:         true,
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: not required

File: turbot/resource_turbot_file.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+func resourceTurbotFileCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	description := d.Get("description")
+	var err error
+	// build input map to pass to mutation
+	input, err := buildFileInput(d, fileProperties)
+	// set type property
+	input["type"] = "tmod:@turbot/turbot#/resource/types/file"
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: why doesn't buildFileInput do this?

Comment (kaidaguerre) [reply turbot/resource_turbot_file.go]: ah because we do not need for update?

File: apiClient/types.go [hunk 3]
```diff
@@ -372,6 +372,7 @@ type TurbotResourceMetadata struct {
 	Metadata          map[string]interface{}
 	Tags              map[string]string
 	Title             string
+	Description       string
```
Comment (kaidaguerre) [inline apiClient/types.go]: why is this here?

**Review by kaidaguerre — CHANGES_REQUESTED** [review]
looks good - couple of minor queries/changes

File: turbot/resource_turbot_file.go (hunk 2 above)
Comment (RupeshPatil20) [reply turbot/resource_turbot_file.go]: Yes because then it will be pass on for Update operation, which is not valid

File: turbot/resource_turbot_file.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+	var err error
+	// build input map to pass to mutation
+	input, err := buildFileInput(d, fileProperties)
+	if err != nil {
+		return err
+	}
+	// set type property
+	input["type"] = "tmod:@turbot/turbot#/resource/types/file"
+	// data should be object - strict
+	// if data( object must ), override the title
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: is this comment needed?

File: turbot/resource_turbot_file.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+	d.Set("title", d.Get("title"))
+	d.Set("description", description)
+	return nil
+}
+
+func resourceTurbotFileRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	id := d.Id()
+
+	// build required properties from data.
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: `	// build required properties from data and metadata`

File: turbot/resource_turbot_file.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+		}
+	}
+
+	if _, ok := d.GetOk("metadata"); ok {
+		var err error = nil
+		metadataConfigProperties, err = helpers.PropertyMapFromJson(d.Get("metadata").(string))
+		if err != nil {
+			return fmt.Errorf("error retrieving properties from resource metadata: %s", err.Error())
+		}
+	}
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]:
```
// pass nil for data as we are not reating any data properties, only the metadata
var properties map[string]string = nil
resource, err := client.ReadResource(id, properties)
```

File: turbot/resource_turbot_file.go [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+import (
+	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-turbot/apiClient"
+	"github.com/terraform-providers/terraform-provider-turbot/helpers"
+	"log"
+)
+
+var fileProperties = []interface{}{"parent", "tags", "akas"}
+var metadataProperties = []interface{}{"title", "description"}
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: these could be a string arrays

File: turbot/resource_turbot_file.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+	data, err := helpers.MapToJsonString(resource.Data)
+	if err != nil {
+		return fmt.Errorf("error building resource data: %s", err.Error())
+	}
+	//metaDataMap is from read operation
+	var metaDataMap = make(map[string]interface{})
+	metaDataMap = resource.Turbot.Custom
+	for _, value := range metadataProperties {
+		// values is an array = ["title","description"]
+		metadataProperty := value.(string)
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: no needed if metadataProperties was a string array

File: turbot/resource_turbot_file.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+	// rebuild data from the resource
+	data, err := helpers.MapToJsonString(resource.Data)
+	if err != nil {
+		return fmt.Errorf("error building resource data: %s", err.Error())
+	}
+	//metaDataMap is from read operation
+	var metaDataMap = make(map[string]interface{})
+	metaDataMap = resource.Turbot.Custom
+	for _, value := range metadataProperties {
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: comment to explain what is metadata properties  - 
```
// metadata properties are the configuration properties that arte stored in the turbot custom metadata
```

File: turbot/resource_turbot_file.go [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+			if configValue, ok := d.GetOk(metadataProperty); ok {
+				d.Set(metadataProperty, configValue)
+			}
+		}
+	}
+	// rebuild metadata from the resource
+	metadata, err := helpers.MapToJsonString(metaDataMap)
+	if err != nil {
+		return fmt.Errorf("error building resource metadata: %s", err.Error())
+	}
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go:124]: this can be removed?

File: turbot/resource_turbot_file.go [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+
+func resourceTurbotFileUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	// build input map to pass to mutation
+	id := d.Id()
+	input, err := buildFileInput(d, getFileUpdateProperties())
+	if err != nil {
+		return err
+	}
+	excludedPropertiesInUpdate, err := client.BuildPropertiesFromUpdateSchema(id, []interface{}{"updateSchema"})
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: we do not need to do this - we know the schema

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: helpers/main.go [hunk 12]
```diff
@@ -143,3 +143,14 @@ func GetNullProperties(propertyMap map[string]interface{}) []string {
 	}
 	return result
 }
+
+//
+func GetOldMapProperties(old, new map[string]interface{}) []interface{} {
```
Comment (kaidaguerre) [inline helpers/main.go:148]: maybe a rename and a comment - what does this function do?

File: turbot/resource_turbot_file.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+	return nil
+}
+
+func resourceTurbotFileRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	id := d.Id()
+
+	// build required properties from content.
+	// properties is a map of property name -> property path
+
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: remove line

File: turbot/resource_turbot_file.go [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+	var properties map[string]string = nil
+
+	if _, ok := d.GetOk("content"); ok {
+		var err error = nil
+		properties, err = helpers.PropertyMapFromJson(d.Get("content").(string))
+		if err != nil {
+			return fmt.Errorf("error retrieving properties from resource data: %s", err.Error())
+		}
+	}
+	// pass nil
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: this comment is wrong I think?

File: turbot/resource_turbot_file.go [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+	d.Set("description", customMetadata["description"])
+	d.Set("content", content)
+	return nil
+}
+
+func resourceTurbotFileUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*apiClient.Client)
+	// build input map to pass to mutation
+	id := d.Id()
+	input, err := buildFileInput(d, getFileUpdateProperties())
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: add comment - what is getFileUpdateProperties doing?

File: turbot/resource_turbot_file.go [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-turbot/apiClient"
+	"github.com/terraform-providers/terraform-provider-turbot/helpers"
+)
+
+var fileProperties = []interface{}{"parent", "tags", "akas"}
+var topLevelProperties = []interface{}{"title", "description"}
+
+func getFileUpdateProperties() []interface{} {
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: Not needed

File: turbot/resource_turbot_file.go (hunk 14 above)
Comment (kaidaguerre) [reply turbot/resource_turbot_file.go]: remove comment

Comment (kaidaguerre) [reply turbot/resource_turbot_file.go]: or change it

File: turbot/resource_turbot_file.go [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+	var input = make(map[string]interface{})
+	var content = make(map[string]interface{})
+	input = mapFromResourceData(d, properties)
+	// convert data from json string to map
+	contentString := d.Get("content").(string)
+	if content, err = helpers.JsonStringToMap(contentString); err != nil {
+		return nil, fmt.Errorf("error build resource mutation input, failed to unmarshal content: \n%s\nerror: %s", contentString, err.Error())
+	}
+	// set old properties of content to NULL in update operation
+	if old, new := d.GetChange("content"); old != nil {
```
Comment (kaidaguerre) [inline turbot/resource_turbot_file.go]: this should be split out from buildFileInput - we only want to do this for an update

File: apiClient/resource.go [hunk 18]
```diff
@@ -39,6 +39,23 @@ func (client *Client) ReadResource(resourceAka string, properties map[string]str
 	return resource, nil
 }
 
+func (client *Client) ReadFullResource(resourceAka string) (*Resource, error) {
+	query := readFullResourceQuery(resourceAka)
```
Comment (kaidaguerre) [inline apiClient/resource.go:43]: does readFullResourceQuery already exist?

Comment (kaidaguerre) [reply apiClient/resource.go:43]: I think we should use `data` instead of `object` - then we would not need to remove the turbot property

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: apiClient/resource.go (hunk 18 above)
Comment (RupeshPatil20) [reply apiClient/resource.go:43]: yes readFullResourceQuery existed but not used
